docs/src/proposals/big-transactions.md

```diff
@@ -0,0 +1,186 @@
+# Big Transactions
+
+## Problem
+
+Messages transmitted to Solana validators must not exceed the IPv6 MTU size to
+ensure fast and reliable network transmission of cluster info over UDP.
+Solana's networking stack uses a conservative MTU size of 1280 bytes which,
+after accounting for headers, leaves 1232 bytes for packet data like serialized
+transactions.
+
+Developers building applications on Solana must design their on-chain program
+interfaces within the above transaction size limit constraint. One common
+work-around is to store state temporarily on-chain and consume that state in
+later transactions. This is the approach used by the BPF loader program for
+deploying Solana programs.
+
+However, this workaround doesn't work well when developers compose many on-chain
+programs in a single atomic transaction. With more composition comes more
+account inputs, each of which takes up 32 bytes. There is currently no available
+workaround for increasing the number of accounts used in a single transaction
+since each transaction must list all accounts that it needs to properly lock
+accounts for parallel execution. Therefore the current cap is about 35 accounts
+after accounting for signatures and other transaction metadata.
+
+## Proposed Solution
+
+Introduce a new on-chain account indexing program which stores account address
+mappings and add a new transaction format which supports concise account
+references through the new on-chain account indexes.
+
+### Account Indexing Program
+
+Here we describe a contract-based solution to the problem, whereby a protocol
+developer or end-user can create collections of related accounts on-chain for
+concise use in a transaction's account inputs. This approach is similar to page
+tables used in operating systems to succinctly map virtual addresses to physical
+memory.
+
+After addresses are stored on-chain in an index account, they may be succinctly
+referenced from a transaction using an index rather than a full 32 byte address.
+This will require a new transaction format to make use of these succinct indexes
+as well as runtime handling for looking up and loading accounts from the
+on-chain indexes.
+
+#### State
+
+Once created, index accounts may not be deleted. Stored addresses should be
+append only so that once an address is stored in an index account, it may not be
+removed later.
+
+Since transactions use a u16 offset to look up addresses, index accounts can
+store up to 2^16 addresses each. Anyone may create an index account of any size
+as long as its big enough to store the necessary metadata. In addition to
+stored addresses, index accounts must also track the latest count of stored
+addresses and an authority which must be a present signer for all index
+modifications.
+
+#### Program controlled indexes
+
+If the authority of an index account is controlled by a program, more
+sophisticated indexes could be built with governance features or price curves
+for new index addresses.
+
```

Comment:
I think initially the index should be created and finalized such that it is no longer modifiable after finalization.  Otherwise attacker can change the index and submit transactions using the updated index values preventing the system from taking advantage of any caches.

1. CreateIndex
2. AddPubkey
3. Finalize (set the finalization slot)
Finalization slot + N slot threshold is when the index is considered enabled.

Reply:
> I think initially the index should be created and finalized such that it is no longer modifiable after finalization.

I think an index that grows over time would be much more powerful and flexible. But this conservative approach is probably good enough for now.

> Otherwise attacker can change the index and submit transactions using the updated index values preventing the system from taking advantage of any caches

Yeah great point. I think this could be solved without the finalization requirement though. If indexes are append only and  new additions only take effect in the next slot, we should be ok.

```
struct IndexMeta {
  authority: Pubkey,
  last_update_slot: Slot,
  index_size: u16,
  addresses_added_in_last_update_slot: u16,
}
```

The runtime will only allow lookups to `meta.index_size - meta.addresses_added_in_last_update_slot` while current slot <= last update slot. This way, banks can always have up to date caches for index accounts for the following block.

Reply:
wouldn't we need to keep N indexes in cache, one for each slot?

Reply:
Well, each bank only needs to store the list of pending addresses that were appended in its own slot. When a bank is rooted/squashed, its recorded addresses can get appended a finalized list which can be shared across all forks. If an unfinalized index entry is referenced, we can do a binary search over the unrooted ancestors of the current block to find the address.

Reply:
Yea, that's much more complex then just using the default account cache used already by the bank.  If the set is finalized we don't need to build new per/bank hashmap.

Reply:
That's fair, the finalization approach is much simpler. I changed the proposal to reflect that these indexes can only be used once finalized. They can still be deactivated at epoch boundaries and have funds recovered though to make it economically viable for devs to create new indexes as needed.